colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -557,55 +596,73 @@ def schedule_f(
                 input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
 
         # Here, let input_obj.requires_grad_()
-        tree_map(torch.Tensor.requires_grad_, input_obj)
+        # if input_obj is not None:
+        if not isinstance(input_obj, torch.Tensor):
+            tree_map(require_grad, input_obj)
+
+        # Also requires_grad_ for micro_batch in stage 0 chunk 0 fwd,
+        # tree_map(torch.Tensor.requires_grad_, micro_batch)
 
         # Step2: fwd step
         output_obj = self.forward_step(
             model_chunk=model_chunk,
             model_chunk_id=model_chunk_id,
+            micro_batch=micro_batch,
             input_obj=input_obj,
             criterion=criterion,
             accum_loss=accum_loss,
             outputs=outputs,
         )
+
+        # Step3: deallocate output for bwd b & w; (do not detach output)
+        deallocate_output_obj = tree_map(clone, output_obj)
+        if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+            # We should not deallocate bwd LOSS
+            pass
+        else:
+            # deallocate output
+            tree_map(deallocate, deallocate_output_obj)
```

Comment:
Is this wrong? We clone outputs and then deallocate. What do we get?

Reply:
> Is this wrong? We clone outputs and then deallocate. What do we get?

I made a bad name for it, "deallocate" function is actually to release tensor.data and keep grad_fn; 
<img width="716" alt="1726825336835" src="https://github.com/user-attachments/assets/83525abb-8945-442b-927d-0bdcba08501f">